### PR TITLE
Add opacity/tint to Panelsprite

### PR DESCRIPTION
**Pull request by Wend1go**
Opacity is working

Tint has an issue that I don't understand. The colour value is accepted and I can see it change in the debugger but the visuals remain unchanged. I used the same functionality as in the tiled sprite and since panel sprites consist of those TilingSprites too it _should_ also work there. But it doesn't. 
I don't know what is wrong here. :confused:

### CODE REVIEW

**Comment by 4ian** [comment]
> I used the same functionality as in the tiled sprite and since panel sprites consist of those TilingSprites too it should also work there. But it doesn't.

There is currently an optimization applied on the panel sprite where the whole object is stored as a bitmap to avoid re-rendering 9 objects separately at every frame (though this may need some measurements to see if it's still useful as objects outside of the camera viewport are not rendered).
This is done there:

https://github.com/4ian/GDevelop/blob/d4a6f4d562cb5358a7e5c92bca0648e736d61566/Extensions/PanelSpriteObject/panelspriteruntimeobject-pixi-renderer.js#L41-L47

Calling `this._spritesContainer.cacheAsBitmap = false` to invalidate the cache should be enough. It's done in `_updateSpritesAndTexturesSize` - I was thinking of making a `invalidateCachedBitmap` method for this but this seems overkill.

**Comment by Wend1go** [comment]
Thanks for the tip, it worked. :)

Noticed something about the wrong corners #1052 while playing with the opacity. It looks like there is also some overlapping of the single components composing the panel sprite:
grafik
grafik
(ignore the yellow box in the middle, it is just a sprite added for testing the opacity)

**Comment by 4ian** [comment]
9 patch sprites are rendered by rendering 9 separate textures. Some, the ones on the border, are partially hidden behind the others - which is ok when everything is fully opaque.
Where have you added the opacity? Have you added it to `_spritesContainer` (I think it's the right think to do)? You may have to do it *after* cacheAsBitmap was set to true, so that opacity is only applied to the resulting bitmap, not on every of the 9 sprites separately

**Comment by Wend1go** [comment]
At first I added the opacity to the sprite container but it did not update the texture immediately even though I added ´this._spritesContainer.cacheAsBitmap = false´ there too. After that, I applied the same loop over all the corners like I did with the tint.
But I'll have a look if I can get the opacity via `_spritesContainer` back somehow since it would also be more efficient than iterating over all the corners.

**Comment by Wend1go** [comment]
I cant get it working 100% correct with only adding the alpha to the `_spritesContainer` like that:
```
gdjs.PanelSpriteRuntimeObjectPixiRenderer.prototype.updateOpacity = function() {
    this._spritesContainer.alpha = this._spritesContainer.visible ? this._object.opacity/255 : 0;
}
```
It does seem to work at first, but when I change the tint afterwards the panel sprite immediately goes to full opacity again. Only after updating the opacity again after that, you get the desired result.

**Comment by 4ian** [comment]
>  but when I change the tint afterwards the panel sprite immediately goes to full opacity again

Can you console.log `this._spritesContainer.alpha ` to see if:
* it's still at the proper value
* it's back at 1?

If it's the proper value, can you set it again like this `this._spritesContainer.alpha = this._spritesContainer.alpha`? Sounds weird, but this might be implemented as a get/set property in JS, meaning that calling the = operator will internally (in Pixi.js) call a function, that might fix the issue (this would be a pixi issue then)
If it's back at 1, might also be a pixi issue. In which case, a possible solution would be to store opacity in the renderer (this._opacity, next to this._spritesContainer and other members) and apply again the opacity when needed (`this._spritesContainer.alpha = this._opacity`). 
Also would be good to search if there are bug reports on Pixi about this, or related to cacheAsBitmap which may be the culprit.

**Comment by Wend1go** [comment]
Ok problem solved by storing the opacity in the renderer and assigning the alpha value again in `ensureUpToDate` when cacheAsBitmap is set.
Working now as it should.

File: Extensions/PanelSpriteObject/panelspriteruntimeobject-pixi-renderer.js [hunk 1]
```diff
@@ -40,12 +41,19 @@ gdjs.PanelSpriteRuntimeObjectPixiRenderer.prototype.getRendererObject = function
 
 gdjs.PanelSpriteRuntimeObjectPixiRenderer.prototype.ensureUpToDate = function() {
     if (this._spritesContainer.visible && this._wasRendered) {
-        this._spritesContainer.cacheAsBitmap = true;
+        this._spritesContainer.alpha = this._opacity;
```
Comment (4ian) [inline Extensions/PanelSpriteObject/panelspriteruntimeobject-pixi-renderer.js]: Can we add a comment here explaining why it's needed?
"// Update the alpha of the container to make sure that it's applied. If not done [...]"

The goal is to give the future reader/developer an explanation of what was wrong and how is this helping :)

File: Extensions/PanelSpriteObject/panelspriteruntimeobject-pixi-renderer.js [hunk 2]
```diff
@@ -29,6 +29,7 @@ gdjs.PanelSpriteRuntimeObjectPixiRenderer = function(runtimeObject, runtimeScene
         this._spritesContainer.addChild(this._borderSprites[i]);
     }
 
+    this._opacity = this._spritesContainer.alpha;
```
Comment (4ian) [inline Extensions/PanelSpriteObject/panelspriteruntimeobject-pixi-renderer.js]: nitpicking: let's keep the name "opacity" for the thing that goes between 0 and 255. As this variable will contain something that is equivalent to alpha in pixi, which goes between 0 and 1, let's call it `this._alpha`.

**Comment by 4ian** [comment]
Added a few comments, otherwise looks good to me :) Can you double check that opacity/tint is working in all situations? Want to be extra sure because we had to add a "hack" for this.

**Comment by Wend1go** [comment]
Seems to work as it should. There might be a 1-frame flashing to full opacity when changing the colour but I'm not 100% sure. May also be my imagination.
The corners are still borked when changing the image, but this was not the topic of this PR.

**Comment by 4ian** [comment]
For the corners yeah, I did quickly check the borders but strangely could not find any issue at first sight in the code.

**Comment by Wend1go** [comment]
@4ian do you think we can merge this before beta 75, when #1126 is fixed?
So we would have all the tint/opacity functions for visible objects released together.

**Comment by 4ian** [comment]
Made a few changes, looks good! Merging now :)

**Comment by 4ian** [comment]
Thanks for your work on these objects! :)

**Pull request merged**